boot/freeldr/freeldr/lib/fs/btrfs.c

```diff
@@ -699,18 +699,32 @@ static u64 btrfs_read_extent_reg(struct btrfs_path *path, struct btrfs_file_exte
         physical += extent->offset + offset;
         if (physical & (512 - 1))
         {
+            u32 shift;
+
             /* If somebody tried to do unaligned access */
-            physical -= offset;
-            temp_out = FrLdrTempAlloc(size + offset, TAG_BTRFS_FILE);
 
-            if (!disk_read(physical, temp_out, size + offset))
+            temp_out = FrLdrTempAlloc(512, TAG_BTRFS_FILE);
+
+            if (!disk_read(physical & ~(512 - 1), temp_out, 512))
             {
                 FrLdrTempFree(temp_out, TAG_BTRFS_FILE);
                 return READ_ERROR;
             }
 
-            memcpy(out, temp_out + offset, size);
+            shift = physical & (512 - 1);
```

Comment:
Need/Want a u64-to-u32 cast?

Reply:
Yes, please make sure that doesn't introduce a warning on MSVC builds.